### PR TITLE
Add idempotent related getter and setter to MethodDescriptor

### PR DESCRIPTION
**Pull request by zsurocking**

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
It looks like the Constructor is ExperimentalApi, You can probably break it and add this param in.

**Comment by zsurocking** [comment]
The constructor is used in proto generated code. We have not yet made the decision to propagate this through proto option. So for now, people using our proto generated surface will need this to set idempotency.

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 1]
```diff
@@ -215,6 +216,24 @@ public InputStream streamResponse(RespT response) {
...
+    return idempotent;
+  }
+
+  /**
+   * Set idempotency on this method.
+   *
+   * @param idempotent the idempotency of this method.
+   * @return this MethodDescriptor.
+   */
+  public MethodDescriptor idempotent(boolean idempotent) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Let's name it `withIdempotent`

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 2]
```diff
@@ -215,6 +216,24 @@ public InputStream streamResponse(RespT response) {
   }
 
   /**
+   * Returns whether this method is idempotent.
+   */
+  public boolean idempotent() {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: `@ExperimentalApi` here and below.

File: core/src/main/java/io/grpc/MethodDescriptor.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Oh, you modify the `MethodDescriptor`. How about we make a copy with the modification in place.

Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Done

File: core/src/main/java/io/grpc/MethodDescriptor.java (hunk 2 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Done. PTAL

**Comment by carl-mastrangelo** [comment]
FWIW Some teams determine idempotency based on some property of the call itself rather than the method.  For example, RPCs that are batches of commands may be idempotent if they consist solely of reads (there are reasons for this).  

Have you considered putting this in CallOptions instead?

**Comment by buchgr** [comment]
I assume idempotency will allow for some optimizations. I wonder what those will be? Automatic retry, on failure maybe? What else?

**Comment by lukaszx0** [comment]
Hah! I was looking into this on Friday but haven't pushed anything cause I had to leave for flight. @buchgr yeah, automatic retries mostly, and I was looking into implementing them on Fri. I'm happy to continue working on this if nobody has already started.

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 3]
```diff
@@ -215,6 +216,28 @@ public InputStream streamResponse(RespT response) {
   }
 
   /**
+   * Returns whether this method is idempotent.
+   */
+  @ExperimentalApi
+  public boolean idempotent() {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: `getIdempotent()`, to match other getters.

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 4]
```diff
@@ -53,6 +53,7 @@
   private final String fullMethodName;
   private final Marshaller<ReqT> requestMarshaller;
   private final Marshaller<RespT> responseMarshaller;
+  private boolean idempotent;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: final, and add another argument to the constructor for it.

**Comment by ejona86** [comment]
@zsurocking, please add a test.

@carl-mastrangelo, we want it to be part of the method. Whether a method is idempotent is actually part of the API. The current plan is for users to specify idempotency in the IDL.

@buchgr, the main push for this at the moment from Google's perspective is to be able to use 0 RTT QUIC. We did see it being useful for retries and so were designing for that as well.

**Comment by carl-mastrangelo** [comment]
@ejona86 Would it make sense for the IDL to be the default for the method, but override able by CallOptions?  This might be too much of an API change for now, but if a grpc users asks for per-call idempotency in the future do you think Call Options would be the right place for it?

**Comment by ejona86** [comment]
@carl-mastrangelo, nobody is asking for per-call idempotency, so I'd mainly defer judgement until we do and decide to support it. It could be done on CallOptions, although maybe only to enable idempotency (not to disable it).

**Comment by zsurocking** [comment]
Test added and fixed all comments. PTAL

**Comment by ejona86** [comment]
@zsurocking LGTM

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 5]
```diff
@@ -215,6 +218,26 @@ public InputStream streamResponse(RespT response) {
   }
 
   /**
+   * Returns whether this method is idempotent.
+   */
+  @ExperimentalApi
+  public boolean getIdempotent() {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Doh. I told you wrong. `isIdempotent` would be more appropriate.

Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Done

**Pull request merged**